File: services/repositories/account_repository.py
"""账号仓储。

Why:
- 将账号相关 SQL 从 `services/database.py` 中逐步下沉。
- 通过依赖注入连接工厂与锁，保持对旧 Facade 的兼容性。
"""

from __future__ import annotations

from threading import Lock
from typing import Callable

import sqlite3


ConnectionFactory = Callable[[], sqlite3.Connection]


class AccountRepository:
    """账号数据仓储（纯数据访问，不包含 GUI/业务编排）。"""

    @staticmethod
    def get_all_accounts(connection_factory: ConnectionFactory, db_lock: Lock) -> list[dict]:
        """查询全部账号。"""
        with db_lock:
            conn = connection_factory()
            cursor = conn.cursor()
            cursor.execute("SELECT * FROM accounts")
            rows = cursor.fetchall()
            conn.close()
            return [dict(row) for row in rows]

    @staticmethod
    def get_account_by_email(
        email: str,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> dict | None:
        """按邮箱查询账号。"""
        if not email:
            return None

        with db_lock:
            conn = connection_factory()
            cursor = conn.cursor()
            cursor.execute("SELECT * FROM accounts WHERE email = ?", (email,))
            row = cursor.fetchone()
            conn.close()
            return dict(row) if row else None

    @staticmethod
    def delete_account(email: str, connection_factory: ConnectionFactory, db_lock: Lock) -> bool:
        """按邮箱删除账号。"""
        with db_lock:
            try:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute("DELETE FROM accounts WHERE email = ?", (email,))
                conn.commit()
                deleted = cursor.rowcount > 0
                conn.close()
                return deleted
            except Exception as error:
                print(f"[DB] 删除账号失败: {error}")
                return False

    @staticmethod
    def get_accounts_needing_unlock(connection_factory: ConnectionFactory, db_lock: Lock) -> list[dict]:
        """查询需要解锁 403 的账号。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute(
                    """
                    SELECT * FROM accounts
                    WHERE unlock_status IN ('needs_unlock', 'unlock_failed')
                    AND validation_url IS NOT NULL
                    AND validation_url != ''
                    """
                )
                rows = cursor.fetchall()
                conn.close()
                return [dict(row) for row in rows]
        except Exception as error:
            print(f"[DB ERROR] get_accounts_needing_unlock 失败: {error}")
            return []

    @staticmethod
    def get_available_pro_accounts(connection_factory: ConnectionFactory, db_lock: Lock) -> list[dict]:
        """查询可用于邀请家庭成员的普通 Pro 账号。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute(
                    """
                    SELECT *, (6 - COALESCE(family_member_count, 0)) as available_slots
                    FROM accounts
                    WHERE is_pro = 'yes'
                    AND COALESCE(family_member_count, 0) < 6
                    AND login_status = 'logged_in'
                    AND browser_profile_id IS NOT NULL
                    AND browser_profile_id != ''
                    ORDER BY family_member_count ASC
                    """
                )
                rows = cursor.fetchall()
                conn.close()
                return [dict(row) for row in rows]
        except Exception as error:
            print(f"[DB ERROR] get_available_pro_accounts 失败: {error}")
            return []

    @staticmethod
    def bind_account_to_browser(
        email: str,
        browser_profile_id: str,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> bool:
        """绑定账号到浏览器窗口。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute(
                    "UPDATE accounts SET browser_profile_id = ?, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
                    (browser_profile_id, email),
                )
                conn.commit()
                affected = cursor.rowcount
                conn.close()
                if affected > 0:
                    print(f"[DB] 绑定账号到窗口: {email} -> {browser_profile_id}")
                return affected > 0
        except Exception as error:
            print(f"[DB ERROR] bind_account_to_browser 失败: {error}")
            return False

    @staticmethod
    def get_account_by_browser(
        browser_profile_id: str,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> dict | None:
        """按窗口 ID 查询账号。"""
        if not browser_profile_id:
            return None
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute(
                    "SELECT * FROM accounts WHERE browser_profile_id = ?",
                    (browser_profile_id,),
                )
                row = cursor.fetchone()
                conn.close()
                return dict(row) if row else None
        except Exception as error:
            print(f"[DB ERROR] get_account_by_browser 失败: {error}")
            return None

    @staticmethod
    def get_unbound_accounts(connection_factory: ConnectionFactory, db_lock: Lock) -> list[dict]:
        """查询未绑定窗口的账号。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute("SELECT * FROM accounts WHERE browser_profile_id IS NULL OR browser_profile_id = ''")
                rows = cursor.fetchall()
                conn.close()
                return [dict(row) for row in rows]
        except Exception as error:
            print(f"[DB ERROR] get_unbound_accounts 失败: {error}")
            return []

    @staticmethod
    def update_sub2api_status(
        email: str,
        status: str,
        account_id: int | None,
        session_id: str | None,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> bool:
        """更新 Sub2API 关联状态。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()

                fields = ["sub2api_status = ?", "updated_at = CURRENT_TIMESTAMP"]
                values: list = [status]

                if account_id is not None:
                    fields.append("sub2api_account_id = ?")
                    values.append(account_id)

                if session_id is not None:
                    fields.append("sub2api_session_id = ?")
                    values.append(session_id)

                values.append(email)
                sql = f"UPDATE accounts SET {', '.join(fields)} WHERE email = ?"
                cursor.execute(sql, values)

                conn.commit()
                affected = cursor.rowcount
                conn.close()

                if affected > 0:
                    suffix = f" (account_id={account_id})" if account_id else ""
                    print(f"[DB] 更新 Sub2API 状态: {email} -> {status}{suffix}")
                return affected > 0
        except Exception as error:
            print(f"[DB ERROR] update_sub2api_status 失败: {error}")
            return False

    @staticmethod
    def get_accounts_by_sub2api_status(
        status: str,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> list[dict]:
        """按 Sub2API 状态查询账号。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute("SELECT * FROM accounts WHERE sub2api_status = ?", (status,))
                rows = cursor.fetchall()
                conn.close()
                return [dict(row) for row in rows]
        except Exception as error:
            print(f"[DB ERROR] get_accounts_by_sub2api_status 失败: {error}")
            return []

    @staticmethod
    def update_login_status(
        email: str,
        status: str,
        last_error: str | None,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> bool:
        """更新账号登录状态。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()

                if status == "logged_in":
                    cursor.execute(
                        "UPDATE accounts SET login_status = ?, last_login_at = CURRENT_TIMESTAMP, "
                        "last_error = NULL, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
                        (status, email),
                    )
                elif last_error is not None:
                    cursor.execute(
                        "UPDATE accounts SET login_status = ?, last_error = ?, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
                        (status, last_error, email),
                    )
                else:
                    cursor.execute(
                        "UPDATE accounts SET login_status = ?, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
                        (status, email),
                    )

                conn.commit()
                affected = cursor.rowcount
                conn.close()

                if affected > 0:
                    print(f"[DB] 更新登录状态: {email} -> {status}")
                return affected > 0
        except Exception as error:
            print(f"[DB ERROR] update_login_status 失败: {error}")
            return False

    @staticmethod
    def update_pro_status(
        email: str,
        is_pro: str,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> bool:
        """更新 Pro 状态。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute(
                    "UPDATE accounts SET is_pro = ?, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
                    (is_pro, email),
                )
                conn.commit()
                affected = cursor.rowcount
                conn.close()

                if affected > 0:
                    print(f"[DB] 更新 Pro 状态: {email} -> {is_pro}")
                return affected > 0
        except Exception as error:
            print(f"[DB ERROR] update_pro_status 失败: {error}")
            return False

    @staticmethod
    def get_accounts_by_login_status(
        status: str,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> list[dict]:
        """按登录状态查询账号。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute("SELECT * FROM accounts WHERE login_status = ?", (status,))
                rows = cursor.fetchall()
                conn.close()
                return [dict(row) for row in rows]
        except Exception as error:
            print(f"[DB ERROR] get_accounts_by_login_status 失败: {error}")
            return []

    @staticmethod
    def get_accounts_for_sub2api(connection_factory: ConnectionFactory, db_lock: Lock) -> list[dict]:
        """查询可用于 Sub2API 关联的账号。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute(
                    """
                    SELECT * FROM accounts
                    WHERE login_status = 'logged_in'
                    AND (sub2api_status IS NULL OR sub2api_status = 'not_linked' OR sub2api_status = 'oauth_failed')
                    """
                )
                rows = cursor.fetchall()
                conn.close()
                return [dict(row) for row in rows]
        except Exception as error:
            print(f"[DB ERROR] get_accounts_for_sub2api 失败: {error}")
            return []

    @staticmethod
    def update_unlock_status(
        email: str,
        status: str,
        validation_url: str | None,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> bool:
        """更新 403 解锁状态。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()

                fields = ["unlock_status = ?", "updated_at = CURRENT_TIMESTAMP"]
                values: list = [status]

                if validation_url is not None:
                    fields.append("validation_url = ?")
                    values.append(validation_url)

                values.append(email)
                sql = f"UPDATE accounts SET {', '.join(fields)} WHERE email = ?"
                cursor.execute(sql, values)

                conn.commit()
                affected = cursor.rowcount
                conn.close()

                if affected > 0:
                    print(f"[DB] 更新解锁状态: {email} -> {status}")
                return affected > 0
        except Exception as error:
            print(f"[DB ERROR] update_unlock_status 失败: {error}")
            return False

    @staticmethod
    def get_accounts_by_unlock_status(
        status: str,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> list[dict]:
        """按解锁状态查询账号。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute("SELECT * FROM accounts WHERE unlock_status = ?", (status,))
                rows = cursor.fetchall()
                conn.close()
                return [dict(row) for row in rows]
        except Exception as error:
            print(f"[DB ERROR] get_accounts_by_unlock_status 失败: {error}")
            return []

    @staticmethod
    def update_family_member_count(
        email: str,
        count: int,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> bool:
        """更新家庭成员数量。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute(
                    "UPDATE accounts SET family_member_count = ?, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
                    (count, email),
                )

                conn.commit()
                affected = cursor.rowcount
                conn.close()

                if affected > 0:
                    print(f"[DB] 更新家庭成员数量: {email} -> {count}")
                return affected > 0
        except Exception as error:
            print(f"[DB ERROR] update_family_member_count 失败: {error}")
            return False

    @staticmethod
    def update_family_sharing_enabled(
        email: str,
        status: str,
        connection_factory: ConnectionFactory,
        db_lock: Lock,
    ) -> bool:
        """更新家庭共享开启状态。"""
        if status not in ("unknown", "yes", "no"):
            print(f"[DB ERROR] 无效的 family_sharing_enabled 状态: {status}")
            return False

        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute(
                    "UPDATE accounts SET family_sharing_enabled = ?, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
                    (status, email),
                )
                conn.commit()
                affected = cursor.rowcount
                conn.close()

                if affected > 0:
                    print(f"[DB] 更新家庭共享状态: {email} -> {status}")
                return affected > 0
        except Exception as error:
            print(f"[DB ERROR] update_family_sharing_enabled 失败: {error}")
            return False

    @staticmethod
    def get_pro_accounts_for_sharing(connection_factory: ConnectionFactory, db_lock: Lock) -> list[dict]:
        """查询可开启共享的普通 Pro 账号。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute(
                    """
                    SELECT * FROM accounts
                    WHERE is_pro = 'yes'
                    AND login_status = 'logged_in'
                    AND browser_profile_id IS NOT NULL
                    AND browser_profile_id != ''
                    AND COALESCE(family_sharing_enabled, 'unknown') != 'yes'
                    ORDER BY updated_at DESC
                    """
                )
                rows = cursor.fetchall()
                conn.close()
                return [dict(row) for row in rows]
        except Exception as error:
            print(f"[DB ERROR] get_pro_accounts_for_sharing 失败: {error}")
            return []

    @staticmethod
    def get_family_pro_accounts(connection_factory: ConnectionFactory, db_lock: Lock) -> list[dict]:
        """查询家庭组 Pro 账号。"""
        try:
            with db_lock:
                conn = connection_factory()
                cursor = conn.cursor()
                cursor.execute(
                    """
                    SELECT * FROM accounts
                    WHERE is_pro = 'family_yes'
                    AND login_status = 'logged_in'
                    """
                )
                rows = cursor.fetchall()
                conn.close()
                return [dict(row) for row in rows]
        except Exception as error:
            print(f"[DB ERROR] get_family_pro_accounts 失败: {error}")
            return []

File: desktop/src/db/account-repository.ts
/**
 * 账号仓储（Node 重写）
 * 对标 services/repositories/account_repository.py
 *
 * POC 范围：只实现读取类方法 + 一个写入方法，用于与 Python 对拍。
 */
import type { Db } from "./connection.ts";

export interface AccountRow {
  email: string;
  password: string | null;
  recovery_email: string | null;
  secret_key: string | null;
  status: string | null;
  login_status: string | null;
  is_pro: string | null;
  browser_profile_id: string | null;
  sub2api_status: string | null;
  unlock_status: string | null;
  updated_at: string | null;
  [key: string]: unknown;
}

export class AccountRepository {
  private readonly db: Db;

  constructor(db: Db) {
    this.db = db;
  }

  /** 对标 get_all_accounts() */
  getAllAccounts(): AccountRow[] {
    return this.db.prepare("SELECT * FROM accounts ORDER BY email").all() as AccountRow[];
  }

  /** 对标 get_account_by_email() */
  getAccountByEmail(email: string): AccountRow | null {
    const row = this.db.prepare("SELECT * FROM accounts WHERE email = ?").get(email);
    return (row as AccountRow) ?? null;
  }

  /** 对标 get_accounts_by_status() */
  getAccountsByStatus(status: string): AccountRow[] {
    return this.db
      .prepare("SELECT * FROM accounts WHERE status = ? ORDER BY email")
      .all(status) as AccountRow[];
  }

  /** 对标 get_accounts_by_login_status() */
  getAccountsByLoginStatus(status: string): AccountRow[] {
    return this.db
      .prepare("SELECT * FROM accounts WHERE login_status = ? ORDER BY email")
      .all(status) as AccountRow[];
  }

  /** 对标 get_account_by_browser() */
  getAccountByBrowser(browserProfileId: string): AccountRow | null {
    const row = this.db
      .prepare("SELECT * FROM accounts WHERE browser_profile_id = ?")
      .get(browserProfileId);
    return (row as AccountRow) ?? null;
  }

  /** 对标 get_unbound_accounts() */
  getUnboundAccounts(): AccountRow[] {
    return this.db
      .prepare(
        "SELECT * FROM accounts WHERE browser_profile_id IS NULL OR browser_profile_id = '' ORDER BY email",
      )
      .all() as AccountRow[];
  }

  /** 按状态分组统计，用于快速核对总量 */
  countByStatus(): Record<string, number> {
    const rows = this.db
      .prepare("SELECT status, COUNT(*) AS n FROM accounts GROUP BY status")
      .all() as { status: string | null; n: number }[];
    const out: Record<string, number> = {};
    for (const r of rows) out[r.status ?? "(null)"] = r.n;
    return out;
  }

  /**
   * 插入或更新账号。对标 upsert_account()。
   *
   * 关键语义：已存在的记录**只更新传入的非 null 字段**，
   * 这样调用方可以只改一个字段（如只换 secret_key）而不覆盖其它数据。
   * 新记录插入时 status 默认 "pending"、sheerid_steps 默认 0。
   */
  upsertAccount(fields: {
    email: string;
    password?: string | null;
    recovery_email?: string | null;
    secret_key?: string | null;
    link?: string | null;
    status?: string | null;
    message?: string | null;
    sheerid_steps?: number | null;
    last_failed_step?: string | null;
    last_error?: string | null;
    browser_profile_id?: string | null;
  }): boolean {
    const email = fields.email;
    if (!email) {
      console.error("[DB] upsert_account: email 为空，跳过");
      return false;
    }

    try {
      const exists = this.db.prepare("SELECT email FROM accounts WHERE email = ?").get(email);

      // 字段名映射：入参键 → 库表列
      const columnMap: [keyof typeof fields, string][] = [
        ["password", "password"],
        ["recovery_email", "recovery_email"],
        ["secret_key", "secret_key"],
        ["link", "verification_link"],
        ["status", "status"],
        ["message", "message"],
        ["sheerid_steps", "sheerid_steps"],
        ["last_failed_step", "last_failed_step"],
        ["last_error", "last_error"],
        ["browser_profile_id", "browser_profile_id"],
      ];

      if (exists) {
        const sets: string[] = [];
        const values: (string | number | null)[] = [];
        for (const [key, column] of columnMap) {
          const v = fields[key];
          if (v === undefined || v === null) continue; // 未传即不动
          sets.push(`${column} = ?`);
          // 空串在 last_* 两列按 Python 的行为落成 NULL
          values.push(
            (column === "last_failed_step" || column === "last_error") && v === "" ? null : v,
          );
        }
        if (sets.length > 0) {
          sets.push("updated_at = CURRENT_TIMESTAMP");
          values.push(email);
          this.db.prepare(`UPDATE accounts SET ${sets.join(", ")} WHERE email = ?`).run(...values);
        }
      } else {
        this.db
          .prepare(
            `INSERT INTO accounts (
               email, password, recovery_email, secret_key,
               verification_link, status, message, sheerid_steps,
               last_failed_step, last_error, browser_profile_id
             ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
          )
          .run(
            email,
            fields.password ?? null,
            fields.recovery_email ?? null,
            fields.secret_key ?? null,
            fields.link ?? null,
            fields.status ?? "pending",
            fields.message ?? null,
            fields.sheerid_steps ?? 0,
            fields.last_failed_step ?? null,
            fields.last_error ?? null,
            fields.browser_profile_id ?? null,
          );
      }
      return true;
    } catch (error) {
      console.error(`[DB ERROR] upsert_account 失败，email: ${email}, 错误: ${error}`);
      return false;
    }
  }

  /**
   * 更新登录状态。对标 update_login_status()。
   * 三分支语义：logged_in 会顺带刷新 last_login_at 并清空 last_error；
   * 带 last_error 时写入错误；否则只改状态。均刷新 updated_at。
   */
  updateLoginStatus(email: string, status: string, lastError?: string | null): boolean {
    try {
      let info;
      if (status === "logged_in") {
        info = this.db
          .prepare(
            "UPDATE accounts SET login_status = ?, last_login_at = CURRENT_TIMESTAMP, " +
              "last_error = NULL, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
          )
          .run(status, email);
      } else if (lastError !== undefined && lastError !== null) {
        info = this.db
          .prepare(
            "UPDATE accounts SET login_status = ?, last_error = ?, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
          )
          .run(status, lastError, email);
      } else {
        info = this.db
          .prepare(
            "UPDATE accounts SET login_status = ?, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
          )
          .run(status, email);
      }
      return Number(info.changes ?? 0) > 0;
    } catch (error) {
      console.error(`[DB ERROR] update_login_status 失败: ${error}`);
      return false;
    }
  }

  /**
   * 更新 Pro 状态。对标 update_pro_status()。
   * is_pro 取值：yes / family_yes / no / unknown
   */
  updateProStatus(email: string, isPro: string): boolean {
    try {
      const info = this.db
        .prepare("UPDATE accounts SET is_pro = ?, updated_at = CURRENT_TIMESTAMP WHERE email = ?")
        .run(isPro, email);
      return Number(info.changes ?? 0) > 0;
    } catch (error) {
      console.error(`[DB ERROR] update_pro_status 失败: ${error}`);
      return false;
    }
  }

  /** 更新家庭组共享开关。对标 update_family_sharing_enabled() */
  updateFamilySharingStatus(email: string, enabled: boolean): boolean {
    try {
      const info = this.db
        .prepare(
          "UPDATE accounts SET family_sharing_enabled = ?, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
        )
        .run(enabled ? "yes" : "no", email);
      return Number(info.changes ?? 0) > 0;
    } catch (error) {
      console.error(`[DB ERROR] update_family_sharing_status 失败: ${error}`);
      return false;
    }
  }

  /** 更新家庭成员数。对标 update_family_member_count() */
  updateFamilyMemberCount(email: string, count: number): boolean {
    try {
      const info = this.db
        .prepare(
          "UPDATE accounts SET family_member_count = ?, updated_at = CURRENT_TIMESTAMP WHERE email = ?",
        )
        .run(count, email);
      return Number(info.changes ?? 0) > 0;
    } catch (error) {
      console.error(`[DB ERROR] update_family_member_count 失败: ${error}`);
      return false;
    }
  }

  /**
   * 更新账号会员信息（刷新家庭组信息专用）。对标 update_membership_info()。
   *
   * 一次性覆盖 9 个字段，并把 family_info_refreshed_at / updated_at 刷成 CURRENT_TIMESTAMP。
   * 参数用单个对象承载（字段名保持 snake_case，与 Python 位置参数一一对应）。
   */
  updateMembershipInfo(fields: {
    /** 账号邮箱 */
    email: string;
    /** Pro 状态 (yes/no/family_yes/detection_failed) */
    is_pro: string;
    /** Pro 计划名称 */
    pro_plan_name: string;
    /** 家庭组角色 (manager/member/none/unknown) */
    family_role: string;
    /** 家庭组管理员邮箱 */
    family_manager_email: string;
    /** 是否有家庭组 (yes/no/unknown) */
    has_family_group: string;
    /** 家庭成员数量 */
    family_member_count: number;
    /** 剩余家庭组位置 */
    family_slots_left: number;
    /** 账户所属国家 */
    account_country: string;
    /** 刷新错误信息 */
    error_message: string | null;
  }): boolean {
    try {
      const info = this.db
        .prepare(
          `UPDATE accounts SET
                        is_pro = ?,
                        pro_plan_name = ?,
                        family_role = ?,
                        family_manager_email = ?,
                        has_family_group = ?,
                        family_member_count = ?,
                        family_slots_left = ?,
                        account_country = ?,
                        family_info_refresh_error = ?,
                        family_info_refreshed_at = CURRENT_TIMESTAMP,
                        updated_at = CURRENT_TIMESTAMP
                    WHERE email = ?`,
        )
        .run(
          fields.is_pro,
          fields.pro_plan_name,
          fields.family_role,
          fields.family_manager_email,
          fields.has_family_group,
          fields.family_member_count,
          fields.family_slots_left,
          fields.account_country,
          fields.error_message,
          fields.email,
        );
      const affected = Number(info.changes ?? 0);

      if (affected > 0) {
        console.log(
          `[DB] 更新会员信息: ${fields.email} -> is_pro=${fields.is_pro}, role=${fields.family_role}`,
        );
      }
      return affected > 0;
    } catch (error) {
      console.error(`[DB ERROR] update_membership_info 失败: ${error}`);
      return false;
    }
  }

  /**
   * 更新解锁状态。对标 update_unlock_status()。
   * validation_url 只在传入时才写入（动态拼字段，与 Python 一致）。
   */
  updateUnlockStatus(email: string, status: string, validationUrl?: string | null): boolean {
    try {
      const fields = ["unlock_status = ?", "updated_at = CURRENT_TIMESTAMP"];
      const values: (string | null)[] = [status];
      if (validationUrl !== undefined && validationUrl !== null) {
        fields.push("validation_url = ?");
        values.push(validationUrl);
      }
      values.push(email);
      const info = this.db
        .prepare(`UPDATE accounts SET ${fields.join(", ")} WHERE email = ?`)
        .run(...values);
      return Number(info.changes ?? 0) > 0;
    } catch (error) {
      console.error(`[DB ERROR] update_unlock_status 失败: ${error}`);
      return false;
    }
  }

  /** 更新 Sub2API 关联状态。对标 update_sub2api_status()，可选字段动态拼。 */
  updateSub2apiStatus(
    email: string,
    status: string,
    accountId?: number | null,
    sessionId?: string | null,
  ): boolean {
    try {
      const fields = ["sub2api_status = ?", "updated_at = CURRENT_TIMESTAMP"];
      const values: (string | number | null)[] = [status];
      if (accountId !== undefined && accountId !== null) {
        fields.push("sub2api_account_id = ?");
        values.push(accountId);
      }
      if (sessionId !== undefined && sessionId !== null) {
        fields.push("sub2api_session_id = ?");
        values.push(sessionId);
      }
      values.push(email);
      const info = this.db
        .prepare(`UPDATE accounts SET ${fields.join(", ")} WHERE email = ?`)
        .run(...values);
      return Number(info.changes ?? 0) > 0;
    } catch (error) {
      console.error(`[DB ERROR] update_sub2api_status 失败: ${error}`);
      return false;
    }
  }

  /** 总数 */
  count(): number {
    const row = this.db.prepare("SELECT COUNT(*) AS n FROM accounts").get() as { n: number };
    return row.n;
  }
}
